File: backend/app/api/v1/predictions.py
"""
Predictions API endpoints
Endpoints for ML predictions: price classification and anomaly detection
"""

from fastapi import APIRouter, HTTPException
from app.schemas.prediction import (
    PredioInput,
    PriceClassificationResponse,
    AnomalyDetectionResponse,
    PredictionFullResponse,
)
from app.services.prediction_service import prediction_service
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post("/classify-price", response_model=PriceClassificationResponse)
async def classify_price(predio: PredioInput):
    """
    Classify the price range of a property

    Returns one of: ALTO, BAJO, MEDIO, LUJO
    """
    try:
        result = prediction_service.clasificar_precio(predio.model_dump())
        return PriceClassificationResponse(**result)
    except Exception as e:
        logger.error(f"Error in price classification endpoint: {e}")
        raise HTTPException(
            status_code=500, detail=f"Error al clasificar precio: {str(e)}"
        )


@router.post("/detect-anomaly", response_model=AnomalyDetectionResponse)
async def detect_anomaly(predio: PredioInput):
    """
    Detect if a property transaction is anomalous

    Returns anomaly detection result (possible fraud, money laundering, or data errors)
    """
    try:
        result = prediction_service.detectar_anomalia(predio.model_dump())
        return AnomalyDetectionResponse(**result)
    except Exception as e:
        logger.error(f"Error in anomaly detection endpoint: {e}")
        raise HTTPException(
            status_code=500, detail=f"Error al detectar anomalías: {str(e)}"
        )


@router.post("/full", response_model=PredictionFullResponse)
async def full_prediction(predio: PredioInput):
    """
    Perform complete prediction: price classification + anomaly detection

    Returns both classification and anomaly detection results in a single call
    """
    try:
        result = prediction_service.prediccion_completa(predio.model_dump())
        return PredictionFullResponse(**result)
    except Exception as e:
        logger.error(f"Error in full prediction endpoint: {e}")
        raise HTTPException(
            status_code=500, detail=f"Error al realizar predicción completa: {str(e)}"
        )

File: frontend/src/app/actions/predictions.ts
"use server";

import {
    PredioInput,
    PriceClassificationResponse,
    AnomalyDetectionResponse,
    PredictionFullResponse,
} from "@/types/prediction";

const BACKEND_URL = process.env.BACKEND_URL || "http://127.0.0.1:8000";

/**
 * Classify the price range of a property using ML model
 */
export async function classifyPrice(
    data: PredioInput
): Promise<PriceClassificationResponse | { error: string }> {
    try {
        const response = await fetch(`${BACKEND_URL}/api/v1/predictions/classify-price`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(data),
            // No cache for predictions
            cache: "no-store",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            console.error("Classification error:", errorData);
            return {
                error: errorData.detail || `Error ${response.status}: ${response.statusText}`,
            };
        }

        const result = await response.json();
        return result;
    } catch (error) {
        console.error("Error calling classification API:", error);
        return {
            error: "No se pudo conectar con el servidor. Verifica que el backend esté corriendo.",
        };
    }
}

/**
 * Detect anomalies in a property transaction using ML model
 */
export async function detectAnomaly(
    data: PredioInput
): Promise<AnomalyDetectionResponse | { error: string }> {
    try {
        const response = await fetch(`${BACKEND_URL}/api/v1/predictions/detect-anomaly`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(data),
            cache: "no-store",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            console.error("Anomaly detection error:", errorData);
            return {
                error: errorData.detail || `Error ${response.status}: ${response.statusText}`,
            };
        }

        const result = await response.json();
        return result;
    } catch (error) {
        console.error("Error calling anomaly detection API:", error);
        return {
            error: "No se pudo conectar con el servidor. Verifica que el backend esté corriendo.",
        };
    }
}

/**
 * Perform full prediction: price classification + anomaly detection
 */
export async function fullPrediction(
    data: PredioInput
): Promise<PredictionFullResponse | { error: string }> {
    try {
        const response = await fetch(`${BACKEND_URL}/api/v1/predictions/full`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(data),
            cache: "no-store",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            console.error("Full prediction error:", errorData);
            return {
                error: errorData.detail || `Error ${response.status}: ${response.statusText}`,
            };
        }

        const result = await response.json();
        return result;
    } catch (error) {
        console.error("Error calling full prediction API:", error);
        return {
            error: "No se pudo conectar con el servidor. Verifica que el backend esté corriendo.",
        };
    }
}
